module/executiondatasync/optimistic_syncing/execution_result_query_provider.go

```diff
@@ -0,0 +1,365 @@
+	merged := c
+
+	if incoming.AgreeingExecutors > 0 {
+		merged.AgreeingExecutors = incoming.AgreeingExecutors
+	}
+
+	if len(incoming.RequiredExecutors) > 0 {
+		merged.RequiredExecutors = incoming.RequiredExecutors
+	}
+
+	if incoming.ResultInFork != flow.ZeroID {
+		merged.ResultInFork = incoming.ResultInFork
+	}
+
+	return merged
+}
+
+// DefaultCriteria is the system default criteria for execution result queries.
+var DefaultCriteria = Criteria{
+	AgreeingExecutors: 2,
+}
+
+// Query contains the result of an execution result query.
+// It includes both the execution result and the execution nodes that produced it.
+type Query struct {
+	// ExecutionResult is the execution result for the queried block
+	ExecutionResult *flow.ExecutionResult
+	// ExecutionNodes is the list of execution node identities that produced the result
+	ExecutionNodes flow.IdentitySkeletonList
+}
+
+// ExecutionResultQueryProvider provides execution results and execution nodes based on criteria.
+// It allows querying for execution results by block ID with specific filtering criteria
+// to ensure consistency and reliability of execution results.
+type ExecutionResultQueryProvider interface {
+	// ExecutionResultQuery retrieves execution results and associated execution nodes for a given block ID
+	// based on the provided criteria. It returns a Query containing the execution result and
+	// the execution nodes that produced it.
+	//
+	// Expected errors during normal operations:
+	//   - backend.InsufficientExecutionReceipts - found insufficient receipts for given block ID.
+	//   - All other errors are potential indicators of bugs or corrupted internal state
+	ExecutionResultQuery(blockID flow.Identifier, criteria Criteria) (*Query, error)
+}
+
+var _ ExecutionResultQueryProvider = (*ExecutionResultQueryProviderImpl)(nil)
+
+// ExecutionResultQueryProviderImpl is a container for elements required to retrieve
+// execution results and execution node identities for a given block ID based on specified criteria.
+type ExecutionResultQueryProviderImpl struct {
+	log zerolog.Logger
+
+	executionReceipts storage.ExecutionReceipts
+	state             protocol.State
+
+	preferredENIdentifiers flow.IdentifierList
+	requiredENIdentifiers  flow.IdentifierList
+
+	rootBlockID     flow.Identifier
+	rootBlockResult *flow.ExecutionResult
+
+	baseCriteria Criteria
+}
+
+// NewExecutionResultQueryProviderImpl creates and returns a new instance of
+// ExecutionResultQueryProviderImpl.
+//
+// No errors are expected during normal operations
+func NewExecutionResultQueryProviderImpl(
+	log zerolog.Logger,
+	state protocol.State,
+	executionReceipts storage.ExecutionReceipts,
+	preferredENIdentifiers flow.IdentifierList,
+	requiredENIdentifiers flow.IdentifierList,
+	operatorCriteria Criteria,
+) (*ExecutionResultQueryProviderImpl, error) {
+	// Root block ID and result should not change and could be cached.
+	rootBlock := state.Params().FinalizedRoot()
```

Comment:
This was copied from the previous implementation, but I believe this is incorrect. It should actually be the spork root block. This returns whatever was used to bootstrap the node, which could also be a block after the spork root.

Only the spork root block would have no receipts.


you can get the spork root block like this:
```golang
	sporkRootBlockHeight := state.Params().SporkRootBlockHeight()
	rootBlockID, err := headers.BlockIDByHeight(sporkRootBlockHeight)
	...
```